front/lib/api/workos.ts

```diff
@@ -147,3 +159,200 @@ export function generateWorkOSAdminPortalUrl({
+  );
+}
+
+async function syncAllGroups(
+  auth: Authenticator,
+  {
+    workspace,
+    directory,
+  }: {
+    workspace: WorkspaceType;
+    directory: WorkOSDirectory;
+  }
+): Promise<void> {
+  logger.info(
+    {
+      workspaceId: workspace.sId,
+      directoryId: directory.id,
+    },
+    "[WorkOS] Starting group sync"
+  );
+
+  const workOS = getWorkOS();
+
+  // TODO(2025-05-26 aubin): paginate here.
+  const { data: groups } = await workOS.directorySync.listGroups({
+    directory: directory.id,
+  });
+
+  for (const workOSGroup of groups) {
+    await upsertGroup(auth, { workspace, workOSGroup, directory });
+  }
+
+  logger.info(
+    {
+      workspaceId: workspace.sId,
+      directoryId: directory.id,
+    },
+    "[WorkOS] Group sync completed."
+  );
+}
+
+async function upsertUser({
+  workspace,
+  directory,
+  workOSUser,
+}: {
+  workspace: WorkspaceType;
+  workOSUser: WorkOSUserWithGroups;
+  directory: WorkOSDirectory;
+}): Promise<void> {
+  const localLogger = logger.child({
+    workspaceId: workspace.sId,
+    workOSUserEmail: workOSUser.email,
+    workOSUserId: workOSUser.id,
+    directoryId: directory.id,
+  });
+
+  localLogger.info("[WorkOS] Upserting user.");
+  if (!workOSUser.email) {
+    localLogger.error("[WorkOS] Cannot sync a user without an email.");
+    return;
+  }
+
+  const user = await UserResource.fetchByEmail(workOSUser.email);
+  if (!user) {
+    await UserResource.makeNew({
+      sId: generateRandomModelSId(),
+      auth0Sub: null,
+      workOSId: workOSUser.id,
+      provider: directory.type,
+      // TODO: not sure what the providerId is here, it does not seem to be used.
+      providerId: null,
+      name: workOSUser.lastName ?? workOSUser.email,
+      username: getUserNicknameFromEmail(workOSUser.email),
+      email: sanitizeString(workOSUser.email),
+      firstName: workOSUser.firstName ?? "",
+      lastName: workOSUser.lastName ?? null,
+      imageUrl: null,
+    });
+    localLogger.info("[WorkOS] User successfully created.");
+  } else {
+    await user.updateInfo(
+      getUserNicknameFromEmail(workOSUser.email),
+      workOSUser.firstName || user.firstName,
+      workOSUser.lastName || user.lastName,
+      workOSUser.email
+    );
+    await user.updateWorkOSId({ workOSId: workOSUser.id });
+
+    localLogger.info("[WorkOS] User successfully updated.");
+  }
+}
+
+async function upsertGroup(
+  auth: Authenticator,
+  {
+    workspace,
+    directory,
+    workOSGroup,
+  }: {
+    workspace: WorkspaceType;
+    workOSGroup: WorkOSGroup;
+    directory: WorkOSDirectory;
+  }
+): Promise<void> {
+  const localLogger = logger.child({
+    workspaceId: workspace.sId,
+    workOSGroupId: workOSGroup.id,
+    directoryId: directory.id,
+  });
+
+  localLogger.info("[WorkOS] Upserting group");
+
+  const group = await GroupResource.fetchByWorkOSGroupId(auth, workOSGroup.id);
+
+  if (!group) {
+    await GroupResource.makeNew({
+      name: workOSGroup.name,
```

Comment:
be careful that we have a unique constraint on the name per workspace, we need to handle the case where a group already exist with that name ?